posts/coding-intentionally-in-bash-grains.md

```diff
@@ -0,0 +1,185 @@
+
+The second square has 2 grains on it.  In binary, `0b10`.  The total so far is 3 (or `0b11`).
+
+The third square has 4 (`0b100`) grains on it.  Total so far: 7 (`0b111`).
+
+The fourth square has 8 grains (`0b1000`).  Total so far: 15 (`0b1111`).
+
+---
+
+Can you see the pattern?
+
+Each square represents another binary digit, and adding all of them together just makes a bunch of 1's.
+
+In the student's solution, we could replace the F's with 64 1's (one for each square)!
+
+```bash
+bc <<< "ibase=2;1111111111111111111111111111111111111111111111111111111111111111"
+```
+
+More intentional, because it more closely matches what the problem gives us.  But, with it being so long and with us not speaking robot, a long, essentially uncountable string of 1's is maybe not an improvement.
+
+### Second option: the brute-force calculation
+
+OK, so maybe we abandon the non-decimal numbering systems altogether.  Why don't we make the code match how we would total up the number of grains on a chessboard by hand, by counting the grains on each square?
+
+```bash
+total=0
+current_grains=1
+
+for square in {1..64}; do
+  total=$( bc <<< "$total + $current_grains" )
+  current_grains=$( bc <<< "$current_grains * 2" )
+done
+
+echo "$total"
+```
+
+That's a lot more readable and understandable.  The code clearly shows that the number of squares on the chess board is a driving factor, as well as the doubling effect for each square.  I think this is better than the initial solution.
+
+However.
+
+It's slow.  Looping, adding, and repeatedly calling out to an external command?  They all stack up to a slow-ish run-time.  Now, is that a huge deal?  No.  If you're scripting this in Bash, you've probably already decided you don't have speed constraints.  But could it be better?  Yes.
+
+### Third option: direct calculation
+
+In the math biz (that's what cool math folks call it), we've got a name for the process "take something and then multiply it by the same number a bunch of times": exponents.
+
+In fact, the number of grains on each square can be expressed as:
+
+```bash
+bc <<< "2^($square - 1)"
+# We subtract one from $square because square 1 has 1 (2^0) grains on it.
+```
+
+So, how do we add all these up without iterating?
+
+Let's consider a *smaller* version of the same problem: a chessboard with 5 squares[^3].
+
+The five squares would have the following number of grains:
+
+```txt
+---------------------
+| 1 | 2 | 4 | 8 |16 |
+---------------------
+```
+
+And the total here would be: 1 + 2 + 4 + 8 + 16 = 31.  Hm.  31 doesn't scream anything obvious at me yet.  Let's go a little bigger.
+
+OK, well what about a 6-square chessboard?
+
+```txt
+-------------------------
+| 1 | 2 | 4 | 8 |16 |32 |
+-------------------------
+```
+
+And the sum: 1 + 2 + 4 + 8 + 16 + 32 = 63.  Hmm... I'm actually starting to see a glimmer of a pattern, but we'll do one more just to be sure.
```

Comment:
I really like this section. It could be worth starting with 

```
-------------
| 1 | 2 | 4 | 
-------------
```

because that's the smallest possible square chessboard, but I don't feel strongly about it.

Reply:
Yeah, I decided not to, because it usually takes a few terms into a pattern before the actual pattern shows up clearly.  I wanted to go far enough out that the pattern stood out (powers of 2 above 16 are maybe a little more recognizable) but hopefully small enough to still be approachable.

Reply:
Yepp, that makes sense!